rust/otap-dataflow/crates/otap/src/otlp_exporter.rs

```diff
@@ -114,52 +129,103 @@ impl Exporter<OtapPdata> for OTLPExporter {
-                .accept_compressed(encoding);
-            trace_client = trace_client
-                .send_compressed(encoding)
-                .accept_compressed(encoding);
-        }
+        let compression = self.config.grpc.compression_encoding();
+        let max_in_flight = self.config.max_in_flight.max(1);
 
         // reuse the encoder and the buffer across pdatas
-        let mut logs_encoder = LogsProtoBytesEncoder::new();
-        let mut metrics_encoder = MetricsProtoBytesEncoder::new();
-        let mut traces_encoder = TracesProtoBytesEncoder::new();
-        let mut proto_buffer = ProtoBuffer::new();
+        let mut logs_proto_encoder = LogsProtoBytesEncoder::new();
+        let mut metrics_proto_encoder = MetricsProtoBytesEncoder::new();
+        let mut traces_proto_encoder = TracesProtoBytesEncoder::new();
+
+        let mut logs_proto_buffer = ProtoBuffer::with_capacity(8 * 1024);
+        let mut metrics_proto_buffer = ProtoBuffer::with_capacity(8 * 1024);
+        let mut traces_proto_buffer = ProtoBuffer::with_capacity(8 * 1024);
+
+        let mut grpc_clients = GrpcClientPool::new(max_in_flight, channel, compression);
+        grpc_clients.prepopulate_clients();
 
+        let mut inflight_exports = InFlightExports::new();
+        let mut pending_msg: Option<Message<OtapPdata>> = None;
+
+        // Main loop: 1) finish ready completions, 2) biased wait for either a completion
+        // or the next message, 3) dispatch work while respecting the in-flight budget.
         loop {
-            match msg_chan.recv().await? {
+            if inflight_exports.len() >= max_in_flight && pending_msg.is_some() {
+                if let Some(completed) = inflight_exports.next_completion().await {
+                    let client = finalize_completed_export(
+                        completed,
+                        &effect_handler,
+                        &mut self.pdata_metrics,
+                    )
+                    .await;
+                    grpc_clients.release(client);
+                }
+                continue;
+            }
+
+            while let Some(completed) = inflight_exports.next_completion().now_or_never().flatten() {
+                let client =
+                    finalize_completed_export(completed, &effect_handler, &mut self.pdata_metrics)
+                        .await;
+                grpc_clients.release(client);
+            }
+
+            // Prefer completions if any are ready, otherwise biased select between completion and recv.
+            let msg = if let Some(msg) = pending_msg.take() {
+                msg
+            } else if inflight_exports.is_empty() {
+                msg_chan.recv().await?
+            } else {
+                let completion_fut = inflight_exports.next_completion().fuse();
+                let recv_fut = msg_chan.recv().fuse();
+                futures::pin_mut!(completion_fut, recv_fut);
+
+                futures::select_biased! {
```

Comment:
I am having trouble reading this, other than to make this select_biased magic work you need the other three lines of magic above. I'm trying to decipher how to read "fuse" in relation to the two futures.